edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+// otherwise it will return a map of json
+let json_res = client
+    .query_single_json(query, &("SomeUserName",))
+    .await?
+    .unwrap();
+
+let as_account: Account = serde_json::from_str(&json_res)?;
+```
+
+## Using JSON with the `edgedb(json)` attribute
+
+Adding the `edgedb(json)` attribute on top of `Queryable` allows unpacking a struct from JSON returned from EdgeDB in a single call:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+#[edgedb(json)]
+pub struct JsonQueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let json_queryable_accounts: Vec<JsonQueryableAccount> = client
+    .query("select <json>Account { username, id }", &())
+    .await
+    .unwrap();
+```
+
+This attribute can also be used on an inner struct value that implements `Queryable`:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct InnerJsonQueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+    #[edgedb(json)]
+    pub some_json: HashMap<String, String>,
+}
+``` 
+
+## Transactions
+
+The client also has a `.transaction()` method that allows atomic [transactions](https://www.edgedb.com/docs/edgeql/transactions). Wikipedia has a good example of a transaction and why it would be best done atomically:
+
+```
+An example of an atomic transaction is a monetary transfer from bank account A
+to account B. It consists of two operations, withdrawing the money from 
+account A and saving it to account B. Performing these operations in an atomic
+transaction ensures that the database remains in a consistent state, that is,
+money is neither lost nor created if either of those two operations fails.
+```
+
+A transaction removing 10 cents from one customer's account and placing it in another's would look like this:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct BankCustomer {
+    pub name: String,
+    pub bank_balance: i32,
+}
+// Customer1 has an account with 110 cents in it.
+// Customer2 has an account with 90 cents in it.
+// Customer1 is going to send 10 cents to Customer 2. This will be a transaction
+// because we don't want the case to ever occur - even for a split second -
+// where one account has sent money while the other has not received it yet.
+
+// After the transaction is over, each customer should have 100 cents.
+
+let query = "with customer := (
+    update BankCustomer filter .name = <str>$0
+    set { bank_balance -= 10 }
+    ),
+    select customer {
+      name,
+      bank_balance
+    };";
+
+cloned_client
+    .transaction(|mut conn| async move {
+        let _res_1: BankCustomer = conn.query_required_single(query, &(c1,)).await?;
+        let _res_2: BankCustomer = conn.query_required_single(query, &(&c2,)).await?;
+    })
+    .await?;
+```
+
+Note that atomic transactions can often be achieved with links and [backlinks](https://www.edgedb.com/docs/edgeql/paths#backlinks) instead of transaction operations, which is both more idiomatic to EdgeDB and easier to use.
```

Comment:
With this example it's not. So this wording is a bit out of place. We can make a page with "tips and tricks for transactions", or put some examples of that into "performance/optimization guide" which we plan to do (ask @raddevon for details).

Reply:
The only detail I have at this point regarding the performance and optimization guide is that there _aren't_ many details. 😅  The idea was brought up in our meeting on Wednesday. I know we want to move one recommendation from the main docs — to reverse the direction of a one-to-many relationship so that they can be changed from `multi` links to `single` links — into some kind of performance guide. Beyond that, I don't know much. I suspect I won't be able to devote a ton of energy to it until after the 3.0 release and maybe until after the Cloud release depending on what that demands from me.

Reply:
This is not urgent too. We can probably add a tag and start collecting cases in the issue tracker.

Reply:
I did find your original comment on not always needing a transaction insightful so do want to keep it in for the benefit of others. I think this could work where the tip is moved to the top of the transaction section to ensure it doesn't look like it's insinuating that the transaction example should be done otherwise.

https://github.com/edgedb/edgedb-rust/commit/7000332bce65d07fc1e78ddd9a77e464c2bb03de

Reply:
I like this particular hint too, but honestly I had a long pause here reading and trying to understand the "advanced" note (without sample code!) before getting into the basic transaction example below. I'd vote for stashing this note to be added in a later/other page.